### ISSUE TITLE
[css-text] Syntax for percentage-of-width-of-space for 'word-spacing'

### ISSUE THREAD

**Issue by fantasai** [w3c/csswg-drafts#3232]
We just resolved in https://github.com/w3c/csswg-drafts/issues/2165 to make percentages be relative to the current font size (like ems, but not computing to an absolute length for inheritance). That means we need a different syntax to represent a length proportional to the width of a space (that inherits as a proportion).

**Comment by faceless2** [comment]
An "sp" unit? This always struck me as an obvious gap in the font-related units, and it's no different to calculating "ch" or "ic"

**Comment by frivoal** [comment]
> An "sp" unit? This always struck me as an obvious gap in the font-related units, and it's no different to calculating "ch" or "ic"

That would inherit as a length, and we're looking for something that inherits as a proportion.

An `sp` unit isn't inherently wrong, but it is sufficiently close to the `ch` unit that the demand isn't terribly strong.

**Comment by litherum** [comment]
A new unit is probably overkill for this. Perhaps just a function would work better?

**Comment by fantasai** [comment]
Alternately, just `<number>` could work. I know Tab hates it, but we're doing the same thing in `tab-size` also, it'd be nice to be consistent.

**Comment by litherum** [comment]
I can't think of a good name for a function, so I think I slightly prefer a raw number.

**Comment by tabatkins** [comment]
I'm *really* against infecting more properties with "`<number>` as `<length>`"; it is impossible to calc() with for completely unnecessary technical reasons that authors shouldn't be exposed to.

**Comment by litherum** [comment]
If there was a function, could you calc inside it?

**Comment by tabatkins** [comment]
A function that resolved to a `<length>` could indeed be used in calc().

**Comment by litherum** [comment]
It looks like `<length>`s [require](https://drafts.csswg.org/css-values-4/#dimension) units, but this new value represents a ratio or percentage. 

How would one represent the concept of "Please set my word spacing to 130% of the space width" in a way that is compatible with `calc()`?

Would `word-spacing: scale-space(calc(100% + 30%))` work?

**Comment by tabatkins** [comment]
The calc() problem I'm talking about is that, in a calculation, plain numbers need to be scalars: if a property takes a length, it's always valid to multiply the length by a scalar (`5px * 3`), and never valid to add it to a scalar (`5px + 3`).  But if a number is a non-scalar value (a length), then this becomes impossible to adjudicate. If you start with a `word-spacing: 1.3` and decide you need it to be 1px wider, you can't write `word-spacing: calc(1.3 + 1px)`, because that fails to type-check, and there's no reasonable way around that. You *need* something to indicate that the non-scalar value is a non-scalar - a unit, a percentage sign, a function that we know evaluates to a non-scalar type like `<length>`.

So yeah, if you wanted to use that last option, then `word-spacing: calc(1px + scale-space(130%))` would work, it's just bit verbose.

(I personally think it's completely fine to have more single-use-case units, even or especially ones with longer names to indicate their special-purpose-ness. `word-spacing: calc(1px + 1.3spacewidth)` seems fine to me, and satisfies my objection from above.)

**Comment by litherum** [comment]
`word-spacing: calc(1px + scale-space(130%))` and `word-spacing: scale-space(130%)` seem fine to me.

**Comment by frivoal** [comment]
I'm not too sure why a function is preferable to a unit. Can you explain?

**Comment by tabatkins** [comment]
From Myles' [earlier comment](https://github.com/w3c/csswg-drafts/issues/3232#issuecomment-1027677696):

> A new unit is probably overkill for this. Perhaps just a function would work better?

**Comment by litherum** [comment]
When I said "overkill" I meant "I see no use case for this outside the property in question, so let's scope the feature accordingly."

**Comment by fantasai** [comment]
@litherum Well, there's `tab-size` also, so we have at least two properties that need this syntax.

`tab-size` is currently using bare numbers. I understand Tab's point about how this is bad, but at least it's consistent. Consistently solving the problem for both properties would also be nice, though. I just don't want us to do one thing here and something else in `tab-size`.

**Comment by fantasai** [comment]
Also, as @frivoal mentioned, this needs to inherit as itself and not as a `<length>` or a `<percentage>`, so it's not something that can be folded into the existing `<length-percentage>` structure. We'd need something like `<length-percentage-ratio>`.

**Comment by fantasai** [comment]
I think we have a few options here:

- Just use `<number>`
- Reuse `fr` units, define mathematical combinations with `<length-percentage>` to be allowed in this property.
- Mint a new unit, e.g. `pr` or whatever, define `<length-percentage-proportion>` or whatever
- Define a function like `space(<number>)` which inherits as itself and can be used inside calc() along with length-percentage.

If we're taking any of the new syntax options, we might consider also allowing bare `<number>` to work so that `tab-size` and `word-spacing` can share the same syntactic space for the same operation.

Please note that while in `tab-size` this always refers to the space character, in `word-spacing` it can apply to a variety of different characters, each individually. _In other words, even its used value is not a length._

**Comment by tabatkins** [comment]
Yes, tab-size is currently a number. It's *terrible*, and I'm going to push very strongly against spreading this precedent any further. It's currently contained to exactly two properties and I don't want it going any further when there's no reason for it. In other words, the goodness of (very vague) precedent is overwhelmed by the practical badness of it.

It is *absolutely bad* to have a length-ish value that can't be combined with other length-ish values.  If vague precedent is the only reason for doing this I'm going to strongly object to it.

> Please note that while in tab-size this always refers to the space character, in word-spacing it can apply to a variety of different characters, and is applied as a multipler per character. In other words, even its used value is not a singular length value that can be applied to the whole element.

So this means that the precedent of tab-size is *even less* applicable.

> Mint a new unit, e.g. pr or whatever, define `<length-percentage-proportion>` or whatever

No need for this, we can still call it a `<length>`, because it is. It just resolves at a different time; we already have lengths that resolve at multiple different times, so this isn't anything special in that regard, and the fact that it has to resolve at *actual value* time is just a new wrinkle, not a fundamental difference.

The unit would just be context-sensitive, and probably, uh, resolve to 0 in any other property I guess? Or perhaps resolve relative to `ch`.

**Comment by fantasai** [comment]
> we already have lengths that resolve at multiple different times

No we don't. They all compute to an absolute length.

**Comment by tabatkins** [comment]
No, length-percentages don't becomes lengths until used value time. When a value can be converted to its canonical unit isn't part of the definition of a type.

Once again, my argument is that making a `<length>` that *cannot be used in calc() with other length units* is a mistake that we should not propagate beyond the two places it currently occurs. The two usages of this pattern don't form a popular well-known usage pattern that authors can extrapolate from, so precedent is weak to nil.

**Comment by fantasai** [comment]
`<length-percentage>` and `<length>` are two different types, Tab. My point is, whatever this new thing this is, it's cannot be part of `<length>`.

**Comment by karip** [comment]
You may want to consider how allowing `word-spacing` to take a number value affects SVG.

[SVG1.1](https://www.w3.org/TR/SVG11/text.html#WordSpacingProperty) defines `word-spacing` so that it can accept numbers: "For SVG, if a `<length>` is provided without a unit identifier (e.g., an unqualified number such as 128), the SVG user agent processes the `<length>` as a width value in the current user coordinate system."

[SVG2](https://svgwg.org/svg2-draft/text.html#WordSpacingProperty) redefines the property so that it doesn't allow lengths without a unit identifier. However, [presentation attributes](https://svgwg.org/svg2-draft/types.html#presentation-attribute-css-value) still replace instances of `<length>` in grammar with `[<length> | <number>]`.

Safari, Firefox and Chrome support `word-spacing` without a unit as a presentation attribute. Safari and Chrome also support it as a style property. Here's [a JSFiddle example](https://jsfiddle.net/8c6tpr3L/3/).

**Comment by tabatkins** [comment]
Ah shoot, and word-spacing is on the list of properties that allow `<quirky-length>`, too: <https://drafts.csswg.org/css-values/#deprecated-quirky-length>

Agenda+ to reverse the resolution and just define a unit, then.

**Comment by fantasai** [comment]
Can we use `x`? Then anywhere we need a multiplier we can use e.g. `2x` and that would be reasonably readable.

(The main impediment to this being that certain people decided to alias `x` to `dppx` ...)

**Comment by tabatkins** [comment]
Right, `x` is already claimed as a `<resolution>`, so we can't use it.
